### ISSUE TITLE
sqlFile + relativeToChangelogFile fails when using a logicalFilePath

### ISSUE THREAD

**Issue by darkcamper** [liquibase/liquibase#1277]
## Environment
**Liquibase Version**: 4.0.0

**Liquibase Integration & Version**:  CLI

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: SQL SERVER 2019

**Operating System Type & Version**: WIN 10 Pro 2004 (19041.388)

## Description
When a changelog has a 'logicalFilePath' set, execution of 'sqlFile' changes that have the 'relativeToChangelogFile' set to true fail with an IOException that says that the sql file can't be found in a number of different paths. If 'logicalFilePath' is not set, everything works

## Steps To Reproduce
* Have a simple XML changelog that has an 'logicalFilePath' attribute in the root 'databaseChangeLog' node.
* Have inside that changelog a changeset with a 'sqlFile' change that points to an existing file in the same folder than the changelog, and that has 'relativeToChangelogFile="true"'

```xml
<databaseChangeLog
    xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.6.xsd
    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd"
    logicalFilePath="a-logical-file-path">
   

<changeSet id="some-changeset" author="NA">
      <sqlFile endDelimiter="GO"
        path="file.sql"
        relativeToChangelogFile="true"
       />
</changeSet> 
</databaseChangeLog>
```

* Run liquibase update against the db

```
liquibase --driver=com.microsoft.sqlserver.jdbc.SQLServerDriver --url="jdbc:sqlserver://localhost:1433;database=the_database" --changeLogFile=changelog.xml  --username=the_username --password=thepassword --logLevel=debug update
```

## Actual Behavior
Executions fails with an IOException.

```
Unexpected error running Liquibase: java.io.IOException: The file file.sql was not found in
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\access-bridge-64.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\cldrdata.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\dnsns.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\jaccess.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\localedata.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\nashorn.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\sunec.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\sunjce_provider.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\sunmscapi.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\sunpkcs11.jar
    - C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre\lib\ext\zipfs.jar

...

Specifying files by absolute path was removed in Liquibase 4.0. Please use a relative path or add '/' to the classpath parameter.
For more information, please use the --logLevel flag
```

## Expected/Desired Behavior
No error happens and the sql file is executed against the database.

## Additional Context
* This is happening since 4.0.0 beta1. Version 3.10 works fine
* Peeking a little inside the code, it seems that SQLFileChange.openSqlStream is using the filepath of the changeset (which contains the logicalFilePath) instead of the physical path of the containing changelog (which seems to be the behavior of v 3.x)

**Comment by nvoxland** [comment]
Thanks, if it is using the logicalFilePath for looking up the changelog that would be a problem. We'll have to take a look

**Comment by molivasdat** [comment]
Hi @darkcamper .  Yes thanks for writing up this issue.  We will add this to the list of issues we are reviewing.

**Comment by sfxcode** [comment]
what is the status of this issue? We like to use 4.x in Production.

Greetings,

Tom

**Comment by molivasdat** [comment]
HI @sfxcode It should be set for the next couple of releases.  We will try to pull it in earlier.

**Comment by hylkevds** [comment]
I suppose the Target for this issue needs to be updated, it still exists in 4.3.0.

**Comment by molivasdat** [comment]
Hi @hylkevds Yes it does, apologies for that.

**Comment by danashurst** [comment]
Is there a workaround for this for the time being? I'm using 4.3.1 as part of a CI pipeline and a master.yaml file

`databaseChangeLog:
  - include:
      relativeToChangelogFile: true
      file: db.myDatabase.changelog-1.0.mysql.sql
  - include:
      relativeToChangelogFile: true
      file: db.myDatabase.changelog.triggers-1.0.yaml
`

with the appropriate files in the same dir as the master.yaml

**Comment by molivasdat** [comment]
Hi @danashurst Can you open another issue for the .sql and yaml issue? This one occurs when using logicalFilePath and your should work.
And if I understand correctly, you have a master.yaml file that includes a formatted sql file and a yaml file.  There shouldn't be a workaround needed as this should work as is.

**Comment by yps1987** [comment]
Hi @molivasdat ,
are there any news on this? Or is there a workaround for the root cause?

Regards
Yves

**Comment by pierre2113** [comment]
Hi,
  I'm also running into this issue, I use liquibase maven plugin.  I'm trying to upgrade from 3.5.3 to latest 4.3.2.
my apache maven is at 3.6.0.  JDK was 1.8,  downloaded jdk 11 still having issues running liquibase mavent under ubuntu linux environment.
all my legacy xml scripts   changeSet's use logicalFilePath  my sqlFile's use relativeToChangelogFile="true"   my sqlfile path's look like   
<sqlFile path="./some_filename.sql" stripComments="false" splitStatements="true" encoding="utf8"  endDelimiter="\n/" relativeToChangelog="true"/>

my legacy scripts still work with liquibase maven plugin up to 3.10.3.
so far I haven't found any  liquibase maven plugin above 4.x that work without any other changes described below.
but I only tried a few in 4.x

the only workaround i've found is a bit of a nuisance

the command line  i had to add  resources:resources
    mvn resources:resources liquibase:update ...

<sqlFile path="**classpath:/liquibase/db/changelog/V001**/some_filename.sql" stripComments="false" splitStatements="true" encoding="utf8"  endDelimiter="\n/" relativeToChangelog="**false**"/>

under my resources folder  ->  src/main/resources     
full relative path to the  sql file is     ./liquibase/db/changelog/V001

mvn resources:resources  run at project directory level    copies all files  
under 
     src/main/resources   
to 
     target/classes

with relativeToChangelog now set to "false"   and sqlFile path changed to  classpath path format only,  can liquibase maven plugin above 4.x find my files now with this path shown below,  I believe the files are being found  under   target/classes folder.
                                   classpath:/liquibase/db/changelog/V001/some_filename.sql

using linux relative path ./liquibase/db/changelog/V001/some_filename.sql  did not work with  relativeToChangelog set to "false"

So far it seems this is a bug introduced in liquibase maven plugin 4.x and higher.    I hope this work around sheds some light finding the bug.

**Comment by imod** [comment]
This is really a show stopper - we are not able to update liquibase to any newer version > 3.10.3 :(

we have the following folder layout:

```
src/main/resources/db
project/
├─ src/
│  ├─ main/
│  │  ├─ resources/
│  │  │  ├─ db/
│  │  │  │  ├─ changelog/
│  │  │  │  │  ├─ data/
│  │  │  │  │  │  ├─ data_1.sql
│  │  │  │  │  ├─ changes_1.xml
│  │  │  │  ├─ master.xml
```

where `master.xml` includes `changes_1.xml` and `changes_1.xml` references `data_1.sql` with `<sqlFile>`:

`<sqlFile path="data/data_1.sql" relativeToChangelogFile="true" dbms="mysql,mariadb" encoding="UTF-8"/>`

unfortunate this fails with:

```
liquibase.exception.UnexpectedLiquibaseException: java.io.IOException: The file data/data_1.sql was not found in...
```

To make this work when executed by the maven-plugin, I can change the path to `db/changelog/data/data_1.sql` (regardless if  `relativeToChangelogFile` is set to `true` or `false`), but we also call this chagesets with docker and there the paths will not match anymore:

```
docker run --rm $DOCKER_IMAGE --url="$DB_URL" --changeLogFile=/liquibase/changelog/db.changelog-master.xml --contexts="!dev" --username="$DB_USERNAME" --password="$DB_PASSWORD" --logLevel=info update;
```

The docker image is a simple image derived from the official liquibase image:

```
FROM liquibase/liquibase:3.10.x

COPY --chown=liquibase:liquibase /liquibase-extension/myextension.jar /liquibase/lib/myextension.jar
COPY --chown=liquibase:liquibase src/main/resources/db /liquibase/changelog
```

So as I understand this problem is, that `<sqlFile>` does not respect the `relativeToChangelogFile` correctly anymore. If it would build the path correctly, then it should find the path of the `sqlFile`.

**Comment by s13o** [comment]
Actually, Liqubasde changes are based on a ServiceLoader approach and it allows you to overwrite the `<sqlFile>` change to provide a temporary custom fix till it will be officially released.
I did it already in my projects, it works

**Comment by pierre2113** [comment]
I'm not a 9-5 java developer nor am I related with the development of this tool so there few things I wont change and check-in (such as pom.xml library versions), but the fix is straightforward its a few line code change in 1 file, I submitted the PR  but it got held up by a build break.  the change worked in my local environment with liquibase 4.0.0, 4.3.3, I don't see why it would not work also in 4.3.4.  with the change I debuged liquibase-core inside inteli-j with no issues.

here's the PR.    https://github.com/liquibase/liquibase/pull/1798

maybe someone who has better idea how to work around the build breaks they can add this change again in a new PR.

**Comment by r2liquibase** [comment]
Hi, @mariochampion ! @imod was asking about this. What's the status on this issue? Thanks!!!

**Comment by mariochampion** [comment]
Hey @imod (and others) -- we are working on a significant 4.4 release, and this has been sorted into the release after that. 4.4 is nearing dev-complete, but to get uat and test-complete will be a couple of weeks still. 

we definitely hear that classpath, relative file path, and searchpath cause frictions for people so we need to look it holistically. thanks for the patience!

**Comment by antenko** [comment]
Hi!
We are blocked by the same issue and can't upgrade liquibase to 4.x.x from 3.5.x.
When you are going to release patch with this fix?
Thanks!

**Comment by smainz** [comment]
Found the same error in LoadDataChange.
A quick search in IDEA shows, the identical code is used in mltiple places:
![image](https://user-images.githubusercontent.com/15262922/131496169-6661f8a4-70df-45f6-be63-2f5d4b00a859.png)

Should  I open a new issue for that?

**Comment by msimko81** [comment]
The bug is still effective on 4.6.1

**Comment by up185014** [comment]
@nvoxland molivasdat is it fixed if so please suggest solution as we are waiting for it since a year

**Comment by NastyPill** [comment]
Yep, really waiting for fix

**Comment by lalo-mx** [comment]
Is a new issue required for more follow up?

**Comment by b-gyula** [comment]
It's a generic issue since 4.x `load[Update]Data` does not work either since then. All the companies I know (e.g. Lufthansa) stick to version 3.10.x because of this... I don't know why this case was closed... Maybe we should open a ticket separately for each failing change...
There is no workaround even theoretically, since you newer know how is the top changeLog referred.

**Comment by nvoxland** [comment]
Thanks for the report that it's still failing in 4.6.2. This issue was for sqlFile which should be fixed up, so it would be better to open a new issue if loadData/loadUpdateData is still broken.

I created #2281 for that. Could you add details to that issue?

**Comment by smainz** [comment]
My PR #2281 schould have fixed all of this.

**Comment by lalo-mx** [comment]
The error is happening For sqlFile  in 4.6.2
## Environment

**Liquibase Version**: 4.6.2

**Liquibase Integration & Version**:  Spring Framework 5.3.13

**Database Vendor & Version**: SQL Server 2019

**Operating System Type & Version**: Windows 10 64 Bits


## Changeset
 
<changeSet id="CheckSumUpdate" author="User 1" runOnChange="true">
	<preConditions onFail="MARK_RAN">
		<dbms type="mssql" />
	</preConditions>
	<sqlFile path="db/sql/other.sql" relativeToChangelogFile="true" />
</changeSet>

## Error 
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquidbase' defined in class path resource [applicationContext.xml]: Invocation of init method failed; nested exception is liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: java.io.IOException: The file db/sql/other.sql was not found in
    - file:/C:/<directory...>/classes/
    - file:/C:/<directory...>/lib/HikariCP-3.4.2.jar
	...................

**Comment by nvoxland** [comment]
You mean #2060 @smainz ?

**Comment by smainz** [comment]
Sorry, it is #2060

**Comment by nvoxland** [comment]
I tried a variety of setups with 4.6.3 and was not able to reproduce the reported continue problems.

Can you reply to #2281 with a more complete example to help me fix the problem @lalo-mx or @msimko81 or @up185014 or anyone else who is still hitting this?

**Comment by lalo-mx** [comment]
@nvoxland Example with the error [sample-liquidbase-4.6.2.zip](https://github.com/liquibase/liquibase/files/7764067/sample-liquidbase-4.6.2.zip). The error is not happening with 3.8.6.

**Comment by lalo-mx** [comment]
@nvoxland I just realized that you mentioned 4.6.3. Where could I get that version?

**Comment by konikvranik** [comment]
I debugged it and the problem is that `searchPath` in `liquibase.integration.spring.SpringResourceAccessor#openStreams` is created by concat of the source file and relative path of the referenced file, but should be concat of **directory** of the source file and the referenced file.

**Comment by konradczajka** [comment]
I was about two report a similar issue but it may be related to this one.
My problem occurs when I use Liquibase in a Spring Boot project where changelog files are kept in the classpath "root" i.e. `src/main/resources` directory.
Attempting to reference another file path as a relative one to the source file doesn't work.

I managed to locate the issue the same way @konikvranik did:
The problem source is in [SpringResourceAccessor@139](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/integration/spring/SpringResourceAccessor.java#L139):
```
searchPath = relativeTo.replaceFirst("/[^/]+$", "") + "/" + path;
```
The `replaceFirst` was probably meant to remove the filename from the path ( = obtain its directory) but it doesn't work when the source path contains only a filename.

A similar implementation is present in other `ResourceAccessor` implementations e.g. [ClassLoaderResourceAccessor@171](https://github.com/liquibase/liquibase/blob/6fcabdc2b7d2d5618ba2643dc9cd621e746eabea/liquibase-core/src/main/java/liquibase/resource/ClassLoaderResourceAccessor.java#L171) but [additional checks](https://github.com/liquibase/liquibase/blob/6fcabdc2b7d2d5618ba2643dc9cd621e746eabea/liquibase-core/src/main/java/liquibase/resource/ClassLoaderResourceAccessor.java#L160) are performed earlier to properly handle "slash-less" paths.

If using a subdirectory to keep changeset files is not a requirement (I haven't found anything in the documentation) then I'd consider it a bug. A bug which could fixed quite easily by mirroring `ClassLoaderResourceAccessor` checks in `SpringResourceAccessor` (I'm ignoring the fact that perfrming path-related operations using regular expressions is always risky :)).

If any maintainer could confirm that it should be possible to use `src/main/resources` directory to store changeset files in a Spring Boot app (without any subdirectories) that I'm happy to create a seaprate issue with a sample app to reproduce the issue and/or prepare a PR with a fix.

**Comment by nbrugger-tgm** [comment]
Is this still an issue in liquibase 4.8.0 because i seem to experience it. Everything works fine unless i do use `java -jar` to start my app? Any workaround?

**Comment by vipulpatel2103** [comment]
My team has migrated from 3.8.x to 4.8.0 recently due to Blackduck security issues. We also face the same issue. Is this resolved or not and in which version? Kindly update as this is really blocking issue.

**Comment by Islam-Hasan** [comment]
Still getting the same issue in 4.9.1. This issue is a blocker and is not close as far as I can tell from my experience and the previous comments.

**Comment by Islam-Hasan** [comment]
Since this is the issue that most people end finding, I'll link to the other open issue: https://github.com/liquibase/liquibase/issues/2281

Is there a workaround though?
